Metadata stamp

chriseth opened the pull request:
This PR adds a swarm hash of the compilation metadata (which also includes swarm hashes for the source files) to the runtime bytecode of each contract.

Fixes #611.

chriseth committed:
Metadata stamp.

axic commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -626,6 +650,43 @@ CompilerStack::Source const& CompilerStack::source(string const& _sourceName) co
+		meta["settings"]["libraries"][library.first] = "0x" + toHex(library.second.asBytes());
+
+	meta["output"]["abi"] = metadata(_contract, DocumentationType::ABIInterface);
+	meta["output"]["natspec"] = metadata(_contract, DocumentationType::NatspecUser);
```
I would include both natspec documents. Why do we need two documents though and not a single one?

axic commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -626,6 +650,43 @@ CompilerStack::Source const& CompilerStack::source(string const& _sourceName) co
+
+	Json::FastWriter writer;
+	writer.omitEndingLineFeed();
+	return writer.write(meta);
```
This could use the new helper, `dev::jsonCompactPrint`.

chriseth replied:
Yes, I have to rebase.

axic commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -626,6 +650,43 @@ CompilerStack::Source const& CompilerStack::source(string const& _sourceName) co
+	for (auto const& s: m_sources)
+	{
+		solAssert(s.second.scanner, "Scanner not available");
+		meta["sources"][s.first]["swarm"] =
```
I would definitely also include a raw `keccak256` of the input.

chriseth replied:
We can include that in the spec, but I do not really see the full point of that, at least as long as the swarm hash is still extremely similar to the sha3 hash.

axic commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -571,26 +595,26 @@ string CompilerStack::absolutePath(string const& _path, string const& _reference
+	shared_ptr<Compiler> compiler = make_shared<Compiler>(m_optimize, m_optimizeRuns);
 	Contract& compiledContract = m_contracts.at(_contract.name());
+	string onChainMetadata = createOnChainMetadata(compiledContract);
+	compiler->compileContract(_contract, _compiledContracts, dev::swarmHash(onChainMetadata));
```
I would definitely include `keccak256` _and_ the swarm hash.

chriseth replied:
Again, what would be the use of that?

axic replied:
If I acquire the metadata through a different channel (such as a registry, or a block explorer), I can still validate it's hash without the need of understanding a specific version of Swarm's hashing algorithm (which is subject to change, whereas `keccak256` is not).

axic commented:
I think the settings part is so much Solidity specific that either we should move them under a `solidity` node or remove the `language` node.

axic commented on libdevcore/SwarmHash.cpp:
```diff
@@ -38,32 +38,28 @@ h256 swarmHashSimple(bytesConstRef _data, size_t _size)
+	return swarmHashSimple(ref, _length);
+}
+
+h256 dev::swarmHash(string const& _input)
```
I would keep both `bytes` and `string` implementations of `swarmHash`. Also it should be a separate commit and not part of the metadata change.

chriseth replied:
We don't have a use for bytes.

chriseth commented:
Concerning the layout of the settings option: Of course it is solidity specific. The keys that are not solidity specific are `version`, `language` and some parts of `output`. The specification would say that `settings` is optional, but if it is present, its structure depends on the language. Also the `compiler` key is language-specific, but I don't see a reason to introduce an extra hierarchy level for the language-specific parts.

The `compiler` key is used to retrieve a compiler for source verification. Combined with `settings` it is possible to reproduce the compilation process.

We might want to move `compilationTarget` (which is also solidity-specific) into `settings`, but it is probably better for it to be a bit more prominent.

Are there any other items we want to include?

The idea here is that we include everything that is required for recompiling and also some things that are useful without recompiling.

I did not include the source mappings and the AST for example, because if you want to use that, you usually also have a compiler to just re-generate them.

chriseth committed:
Incorporate comments.

chriseth commented:
Updated.

chriseth commented:
Agreed to use `userdoc` and `devdoc` instead of `natspec`

chriseth commented:
`compilationTarget` should move into settings

axic commented:
Is https://github.com/ethereum/wiki/wiki/NatSpec-Determination superseded by this metadata proposal?

chriseth commented:
Yes, the wiki page is superseded.

chriseth committed:
Make sure some keys are present.

chriseth committed:
Documentation.

chriseth commented:
@axic @pirapira please review, this would be my final version for now.

chriseth committed:
Changelog entry.

pirapira commented on docs/miscellaneous.rst:
```diff
@@ -221,6 +221,125 @@ Either add ``--libraries "Math:0x12345678901234567890 Heap:0xabcdef0123456"`` to
+        keccak256: "0x123..."
+      },
+      // Required: Compilation source files/source units, keys are file names
+      sources:
```
Can the ordering of input files affect the outcome?

chriseth replied:
No, they are sorted by name in the compiler and also here, but good point!

pirapira replied:
OK.  Other compilers can add their own fields indicating the order of sources.

pirapira commented on docs/miscellaneous.rst:
```diff
@@ -221,6 +221,125 @@ Either add ``--libraries "Math:0x12345678901234567890 Heap:0xabcdef0123456"`` to
+      // Required: Compiler settings
+      settings:
+      {
+        // Required for Solidity: Sorted list of remappings
```
Command line arguments in `argv` would be useful for any compiler (this can be deferred to version 2 of this format).

chriseth replied:
We are trying to go away from restricting us to a commandline interface for compiler input. Instead, we want to go to a json object for the settings, which would be compatible with this one.

pirapira replied:
I see.

pirapira commented on docs/miscellaneous.rst:
```diff
@@ -221,6 +221,125 @@ Either add ``--libraries "Math:0x12345678901234567890 Heap:0xabcdef0123456"`` to
+
+Specifically, the runtime code for a contract always starts with
+``push32 <metadata hash> pop``, so you can take a look at the 32 bytes starting at
+the second byte of the code of a contract.
```
Can this be put in the last 33 bytes of the runtime code?  Then we save some runtime gas.

chriseth replied:
I have thought about that, too. Currently, this costs 5 gas, which is minimal but might add up to quite a bit.
I will take a look and check if that is doable.

pirapira commented on libdevcore/SwarmHash.h:
```diff
@@ -20,12 +20,13 @@
 #include <libdevcore/FixedHash.h>
-#include <libdevcore/Common.h>
+
+#include <string>
```
nitpick: the coding rule says STL includes should be placed above `libdevcore/*`.

chriseth replied:
Let's change the coding rule: http://stackoverflow.com/a/614333/5197297 (read the full answer)

pirapira replied:
OK.  I support the change.

pirapira commented on libsolidity/ast/AST.cpp:
```diff
@@ -63,6 +63,14 @@ SourceUnitAnnotation& SourceUnit::annotation() const
+	ASTNode const* scope = m_scope;
+	while (dynamic_cast<Declaration const*>(scope) && dynamic_cast<Declaration const*>(scope)->m_scope)
+		scope = dynamic_cast<Declaration const*>(scope)->m_scope;
+	return dynamic_cast<SourceUnit const&>(*scope).annotation().path;
```
`scope` here might be `nullptr` when `m_scope` is `nullptr`.

pirapira commented on libsolidity/codegen/CompilerContext.cpp:
```diff
@@ -227,6 +227,13 @@ void CompilerContext::injectVersionStampIntoSub(size_t _subIndex)
+{
+	eth::Assembly& sub = m_asm->sub(_subIndex);
+	sub.injectStart(Instruction::POP);
+	sub.injectStart(u256(_metadataHash));
```
When the number `_metadataHash` is accidentially small, this could produce `PUSH31` instead of `PUSH32`.

chriseth replied:
Good find! It might be quite hard to force that to `push32`. Would it be better to change the documentation/specification?
On the other hand, if we add that to the end of the bytecode, then there might be another way to do it.

pirapira replied:
Either way is fine: changing the documentation or forcing `PUSH32`.

pirapira replied:
I saw the change.  It looks good.

pirapira commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -356,32 +367,47 @@ Json::Value const& CompilerStack::metadata(string const& _contractName, Document
 	// caches the result
 	if (!*doc)
-		doc->reset(new Json::Value(InterfaceHandler::documentation(*currentContract.contract, _type)));
+		doc->reset(new Json::Value(InterfaceHandler::documentation(*_contract.contract, _type)));
```
`_contract.contract` could be a `nullptr`.

chriseth committed:
Add swarm hash to the end of the bytecode.

chriseth committed:
Add test.

chriseth commented:
Metadata is now appended to the end of the runtime code. All other comments are also addressed.

chriseth committed:
Fix tests.

axic commented:
My comments on the other PR relating to this are still not addressed 😉

chriseth commented:
@axic sorry, I could not find your comments. Can you please repeat them here?

chriseth committed:
Fix documentation error.

chriseth committed:
Fix emscripten build.

chriseth commented:
Proposal: add magic number that specifies the encoding of the metadata hash. Options include `bzzr0` - should ask swarm people about how they want to version their hash. We could also cbor-encode the magic number and the hash.

chriseth commented:
@axic CBOR encoding of `["bzzr","abtttthhhhhhhhhhhhhhtttttttttttc"]` and `{"bzzr","abtttthhhhhhhhhhhhhhtttttttttttc"}` both need 40 bytes, so 4 bytes more than the actual payload. The problem with cbor is also that we need to find the start of the cbor encoding - not sure if that is doable at all. So I think we have to find our own modified encoding anyway.

If we want to stay flexible, I see some options:

 - reverse CBOR
 - last byte is length (where values larger than 127 are reselved), up to that, cbor encoded
 - last byte is length (...), up to that magic-number prefix plus data

axic commented:
> The problem with cbor is also that we need to find the start of the cbor encoding

That is a valid concern and I would agree the easiest is to leave a trailing fixed-length size field. I would argue to make it 16 bit wide.

Regarding the actual CBOR content I was thinking about an array, starting with a flag byte (integer):
- bit 0 could mean "Swarm hash of metadata"
- bit 1 could mean "keccak256 hash of metadata"

axic commented:
As discussed, a good solution currently seems to:
- place CBOR encoded data immediately following the bytecode
- use a CBOR hashmap
- the `bzzr0` key has the current Swarm hash as the value
- the CBOR is followed by a 16 bit big endian length field, representing the total size of the CBOR encoded data, excluding this length field

chriseth committed:
Use CBOR encoding.

axic commented on docs/miscellaneous.rst:
```diff
@@ -319,13 +321,29 @@ Comments are of course also not permitted and used here only for explanatory pur
+encoding. The current version of the Solidity compiler thus adds the following
+to the end of the deployed bytecode::
+
+    0xa1 0x65 'b' 'z' 'z' 'r' '0' 0x58 0x20 <32 bytes swarm hash> 0x00 0x29
```
0x29 -> 0x1d as the length is 29 in decimal

axic committed:
Update metadata documentation

axic commented on libsolidity/codegen/CompilerContext.cpp:
```diff
@@ -220,13 +220,6 @@ void CompilerContext::appendInlineAssembly(
-{
-	eth::Assembly& sub = m_asm->sub(_subIndex);
-	sub.injectStart(Instruction::POP);
-	sub.injectStart(fromBigEndian<u256>(binaryVersion()));
```
Should we include this in the CBOR?

chriseth replied:
As we said during the call, let's do it once it is necessary.

chriseth committed:
Fix tests

axic commented on test/libsolidity/SolidityABIJSON.cpp:
```diff
@@ -731,6 +734,23 @@ BOOST_AUTO_TEST_CASE(function_type)
+	bytes hash = dev::swarmHash(m_compilerStack.onChainMetadata("test")).asBytes();
+	BOOST_REQUIRE(hash.size() == 32);
+	BOOST_REQUIRE(bytecode.size() >= hash.size());
+	BOOST_CHECK(std::equal(hash.begin(), hash.end(), bytecode.end() - 32));
```
This needs to be updated for CBOR.

axic commented on libdevcore/SwarmHash.cpp:
```diff
@@ -38,13 +38,14 @@ h256 swarmHashSimple(bytesConstRef _data, size_t _size)
 }
 
-h256 swarmHashIntermediate(bytes const& _input, size_t _offset, size_t _length)
+h256 swarmHashIntermediate(string const& _input, size_t _offset, size_t _length)
```
I would still like to keep both `bytes` and `string` options. I think it will be useful to hash other kind of data.

chriseth replied:
Again, let's add the other option back when we need it, currently we only need this one function.

axic replied:
Wondering why is it better changing the code back and forth in that case? 😃

chriseth commented:
@axic added a PR to fix the tests.

axic commented:
We have a test failure:

> /home/travis/build/ethereum/solidity/test/libsolidity/SolidityOptimizer.cpp(75): error in "unused_expressions": Optimizer did not reduce bytecode size.
> /home/travis/build/ethereum/solidity/test/libsolidity/SolidityOptimizer.cpp(75): error in "dead_code_elimination_across_assemblies": Optimizer did not reduce bytecode size.

axic committed:
Fix documentation errors

Subheadings must use consistent format across a file.

The metadata with comments cannot be parsed as JSON.

axic commented:
Locally for some reason this isn't compiling:
```
libsolidity/interface/CompilerStack.cpp:691:42: error: conversion from 'const size_t' (aka 'const unsigned long') to 'Json::Value' is ambiguous
        meta["settings"]["optimizer"]["runs"] = m_optimizeRuns;
                                                ^~~~~~~~~~~~~~
```

axic committed:
Fix type of optimizeRuns

axic committed:
Optimizer tests to not interfere with the optimiser flag

axic committed:
Remove unneeded optimised test from EndToEndTest

chriseth commented:
Result of investigation on my local machine: I can reproduce the test failure with clang, but if I compile the contract manually with optimization with clang and gcc, I get the same result.
This could mean that the bug is in the test framework.

axic commented:
Since a `STOP` is emitted before the aux data, should we mention that in the hash description? e.g. the CBOR data should be preceded by `0`.

axic commented on libsolidity/interface/CompilerStack.h:
```diff
@@ -113,14 +113,15 @@ class CompilerStack: boost::noncopyable
+	);
 	/// Parses and compiles the given source code.
 	/// @returns false on error.
 	bool compile(std::string const& _sourceCode, bool _optimize = false);
```
Shouldn't this also have the `_runs` parameter? Is there a need two support both APIs (this and the preceding one)?

axic commented on libsolidity/interface/CompilerStack.h:
```diff
@@ -252,6 +257,9 @@ class CompilerStack: boost::noncopyable
 
 	ReadFileCallback m_readFile;
+	bool m_optimize = false;
+	unsigned m_optimizeRuns = 200;
```
Probably these two shouldn't have a default here as they are set in `compile`?

chriseth replied:
I try to always set a default value for PODs to keep determinism.

axic reviewed (changes requested):
There should be a test for the content of the metadata JSON. The test contract should also use a library.

axic committed:
Submit library addresses to the compiler in tests (and not link afterwards)

chriseth commented:
@axic I think verifying the content of the metadata output can be left for the pull request that introduces the input format.

chriseth committed:
Add the `_runs` parameter.

chriseth commented:
Concerning the STOP: We cannot rely on the fact that there is no other data before the metadata cbor.

chriseth committed:
Count instructions without metadata hash.

chriseth commented:
BUILDS ARE GREEEN! *jump*
